# 데이터 불러오기

In [1]:
from matplotlib import rc
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import os
import pickle

import warnings
warnings.filterwarnings('ignore')

In [2]:
rc('font', family='AppleGothic')
plt.rcParams['axes.unicode_minus'] = False

# 경쟁사이용
competitor = pd.read_csv('../../data/경쟁사이용.txt', encoding='cp949')
display(competitor)
# 고객DEMO
pp_demo = pd.read_pickle("../../DATA_preprocessed/pp_demo.pickle")
display(pp_demo)
# 멤버심 여부
member = pd.read_csv('../../data/멤버십여부.txt', encoding='cp949')
display(member)
# 채널이용
channel = pd.read_csv('../../data/채널이용.txt', encoding='cp949')
display(channel)

# 상품 분류
pp_prodcl = pd.read_pickle("../../DATA_preprocessed/pp_prodcl.pickle")
display(pp_prodcl)

# 상금 거래 관련
pp_purprd_prodcl = pd.read_pickle('../../DATA_preprocessed/pp_purprd_prodcl.pickle')
pp_purprd_prodcl

고객번호 제휴사  경쟁사    이용년월
0          2   D  D02  201507
1         51   D  D01  201504
2         77   D  D02  201503
3         77   D  D02  201506
4         77   D  D02  201507
...      ...  ..  ...     ...
28154  19375   B  B01  201501
28155  19375   B  B01  201507
28156  19375   C  C03  201511
28157  19376   B  B01  201506
28158  19377   B  B01  201504

[28159 rows x 4 columns]

고객번호 성별      연령대       거주지역
0          1  M    60세이상  서울특별시 강남구
1          2  M    60세이상        경기도
2          3  M    60세이상  서울특별시 은평구
3          4  F    60세이상  서울특별시 노원구
4          5  M    60세이상        경기도
...      ... ..      ...        ...
19378  19379  F  20세~24세  서울특별시 동작구
19379  19380  F  20세~24세  서울특별시 마포구
19380  19381  F  40세~44세      부산광역시
19381  19382  F  45세~49세      부산광역시
19382  19383  F  25세~29세       경상남도

[19383 rows x 4 columns]

고객번호  멤버십명    가입년월
0        11  하이마트  201512
1        21  하이마트  201506
2        37  하이마트  201306
3        43  하이마트  201403
4        44  하이마트  201411
...     ...   ...     ...
7451  19377    롭스  201411
7452  19379    롭스  201412
7453  19380    롭스  201412
7454  19381   다둥이  201506
7455  19383    롭스  201507

[7456 rows x 3 columns]

고객번호           제휴사  이용횟수
0         7  A_MOBILE/APP     4
1        14  A_MOBILE/APP     1
2        42  B_MOBILE/APP    23
3        74  A_MOBILE/APP     1
4        74  B_MOBILE/APP    30
...     ...           ...   ...
8819  19168  C_ONLINEMALL    18
8820  19291  C_ONLINEMALL     1
8821  19312  B_ONLINEMALL   137
8822  19351  C_ONLINEMALL    19
8823  19356  C_ONLINEMALL    16

[8824 rows x 3 columns]

대분류명 대중분류명 제휴사    중분류명 제휴사_중분류명    소분류코드      소분류명 TYPE lev
0     일상용품  생활용품   A    일용잡화    A일용잡화  A010101      위생세제  편의품  중가
1     일상용품  생활용품   A    일용잡화    A일용잡화  A010102       휴지류  편의품  저가
2     일상용품  생활용품   A    일용잡화    A일용잡화  A010103      뷰티상품  편의품  중가
3     일상용품  생활용품   A    일용잡화    A일용잡화  A010104      일용잡화  편의품  중가
4     일상용품  생활용품   A    일용잡화    A일용잡화  A010105      세트상품  편의품  고가
...    ...   ...  ..     ...      ...      ...       ...  ...  ..
4381  일상용품   화장품   D  아이메이크업  D아이메이크업  D080204     아이섀도우  편의품  저가
4382  일상용품   화장품   D  아이메이크업  D아이메이크업  D080205    메이크업세트  편의품  저가
4383  일상용품   화장품   D   립메이크업   D립메이크업  D080301  립스틱/립라이너  편의품  저가
4384  일상용품   화장품   D   립메이크업   D립메이크업  D080302   립글로즈/틴트  편의품  저가
4385  일상용품  케어용품   D  네일메이크업  D네일메이크업  D080401  일반네일/케어류  편의품  중가

[4386 rows x 9 columns]

고객번호 제휴사    영수증번호  점포코드  구매_연  구매_월  구매_일  구매시간    소분류코드   구매금액  \
0             1   A  1182697    12  2014     1     8    14  A010302   4800   
1             1   A  1239229    12  2014     1    18    15  A010302   3500   
2             1   A  1307657    12  2014     1    30    11  A010302   3900   
3             1   A  1402358    12  2014     2    18    16  A010302   3500   
4             1   A  1456379    12  2014     2    28    15  A010302   3500   
...         ...  ..      ...   ...   ...   ...   ...   ...      ...    ...   
28437049  18697   C   631324    29  2014     5    30    21  C060405   2300   
28437050  18697   C   631629    29  2014     7    20    15  C060405   2300   
28437051  18710   C  1073027    75  2015     1    22    12  C030312   7990   
28437052  18710   C  1073138    75  2015     1    29    21  C030312   7990   
28437053  19364   D   209812    13  2015     7    11    21  D040112  10500   

         대분류명 대중분류명  중분류명 제휴사_중분류명    소분류명 TYPE lev  
0          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  
1          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  
2          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  
3          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  
4          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  
...       ...   ...   ...      ...     ...  ...  ..  
28437049   식품  즉석식품  베이커리    C베이커리   페스츄리류  편의품  중가  
28437050   식품  즉석식품  베이커리    C베이커리   페스츄리류  편의품  중가  
28437051   식품   채소류   근채류     C근채류  기타한방약재  편의품  고가  
28437052   식품   채소류   근채류     C근채류  기타한방약재  편의품  고가  
28437053   식품  공산식품    과자      D과자  컵아이스크림  편의품  중가  

[28437054 rows x 17 columns]

# EDA

### pp_purprd_prodcl : 분기 Feature 만들기

In [3]:

cond_2014_1 =(pp_purprd_prodcl['구매_연']==2014)&(pp_purprd_prodcl['구매_월']<=3)
cond_2014_2 =(pp_purprd_prodcl['구매_연']==2014)&((pp_purprd_prodcl['구매_월']>3)&(pp_purprd_prodcl['구매_월']<=6))
cond_2014_3 =(pp_purprd_prodcl['구매_연']==2014)&((pp_purprd_prodcl['구매_월']>6)&(pp_purprd_prodcl['구매_월']<=9))
cond_2014_4 =(pp_purprd_prodcl['구매_연']==2014)&((pp_purprd_prodcl['구매_월']>9)&(pp_purprd_prodcl['구매_월']<=12))
cond_2015_1 =(pp_purprd_prodcl['구매_연']==2015)&(pp_purprd_prodcl['구매_월']<=3)
cond_2015_2 =(pp_purprd_prodcl['구매_연']==2015)&((pp_purprd_prodcl['구매_월']>3)&(pp_purprd_prodcl['구매_월']<=6))
cond_2015_3 =(pp_purprd_prodcl['구매_연']==2015)&((pp_purprd_prodcl['구매_월']>6)&(pp_purprd_prodcl['구매_월']<=9))
cond_2015_4 =(pp_purprd_prodcl['구매_연']==2015)&((pp_purprd_prodcl['구매_월']>9)&(pp_purprd_prodcl['구매_월']<=12))

pp_purprd_prodcl['분기']=0

pp_purprd_prodcl.loc[cond_2014_1,'분기']='14_1'
pp_purprd_prodcl.loc[cond_2014_2,'분기']='14_2'
pp_purprd_prodcl.loc[cond_2014_3,'분기']='14_3'
pp_purprd_prodcl.loc[cond_2014_4,'분기']='14_4'
pp_purprd_prodcl.loc[cond_2015_1,'분기']='15_1'
pp_purprd_prodcl.loc[cond_2015_2,'분기']='15_2'
pp_purprd_prodcl.loc[cond_2015_3,'분기']='15_3'
pp_purprd_prodcl.loc[cond_2015_4,'분기']='15_4'

pp_purprd_prodcl

고객번호 제휴사    영수증번호  점포코드  구매_연  구매_월  구매_일  구매시간    소분류코드   구매금액  \
0             1   A  1182697    12  2014     1     8    14  A010302   4800   
1             1   A  1239229    12  2014     1    18    15  A010302   3500   
2             1   A  1307657    12  2014     1    30    11  A010302   3900   
3             1   A  1402358    12  2014     2    18    16  A010302   3500   
4             1   A  1456379    12  2014     2    28    15  A010302   3500   
...         ...  ..      ...   ...   ...   ...   ...   ...      ...    ...   
28437049  18697   C   631324    29  2014     5    30    21  C060405   2300   
28437050  18697   C   631629    29  2014     7    20    15  C060405   2300   
28437051  18710   C  1073027    75  2015     1    22    12  C030312   7990   
28437052  18710   C  1073138    75  2015     1    29    21  C030312   7990   
28437053  19364   D   209812    13  2015     7    11    21  D040112  10500   

         대분류명 대중분류명  중분류명 제휴사_중분류명    소분류명 TYPE lev    분기  
0          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  
1          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  
2          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  
3          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  
4          식품   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  
...       ...   ...   ...      ...     ...  ...  ..   ...  
28437049   식품  즉석식품  베이커리    C베이커리   페스츄리류  편의품  중가  14_2  
28437050   식품  즉석식품  베이커리    C베이커리   페스츄리류  편의품  중가  14_3  
28437051   식품   채소류   근채류     C근채류  기타한방약재  편의품  고가  15_1  
28437052   식품   채소류   근채류     C근채류  기타한방약재  편의품  고가  15_1  
28437053   식품  공산식품    과자      D과자  컵아이스크림  편의품  중가  15_3  

[28437054 rows x 18 columns]

### 분석할 Feature Set
- {대분류명, TYPE, lev, 구매금액} 중 매출 비중이 높은 카테고리를 선정

In [4]:
feature_set = {
    '대분류명' : ['패션잡화','식품','일상용품'],
    'TYPE' : ['편의품','선매품'],
    'lev' : ['중가','고가'],
    '구매금액' : ['구매금액']
}

### 구매 금액 등급화 및 변동지수(target data)생성

In [5]:
# 등급 부여 사용자 정의 함수
def make_cust_sale_grade(x, y):
    if x > fractile_list[y][7]:
        return 9
    elif x > fractile_list[y][6]:
        return 8
    elif x > fractile_list[y][5]:
        return 7
    elif x > fractile_list[y][4]:
        return 6
    elif x > fractile_list[y][3]:
        return 5
    elif x > fractile_list[y][2]:
        return 4
    elif x > fractile_list[y][1]:
        return 3
    elif x > fractile_list[y][0]:
        return 2
    else:
        return 1 
    
# 고객번호만 들어간 빈 DataFrame 생성
pp_datasets = pp_demo['고객번호']

for i in feature_set.keys():
    for j in range(len(feature_set[i])):
        # 분석을 위한 카테고리별 pivot table 생성
        option = 'sum'
        if feature_set[i][j] == '구매금액':
            pp_pivot = pp_purprd_prodcl.pivot_table(index=['고객번호'], columns=['분기'], values='구매금액', aggfunc=(f'{option}'), fill_value=0)
        else:
            pp_pivot = pp_purprd_prodcl[pp_purprd_prodcl[i]==feature_set[i][j]].pivot_table(index=['고객번호'], columns=['분기'], values='구매금액', aggfunc=(f'{option}'), fill_value=0)
        
        # 등급 분위수 딕셔너리 생성
        Q_list = ['14_1', '14_2', '14_3', '14_4', '15_1', '15_2', '15_3', '15_4']
        fractile_list = {}
        for Q in Q_list:
            fractiles = []
            for percentile in [4, 11, 23, 40, 60, 77, 89, 96]:
                fractiles.append(np.percentile(pp_pivot[Q], percentile))
                fractile_list[Q] = fractiles
                
        # 고개별 분기별 feature 별 매출 등급 생성        
        for Q in Q_list:
            pp_pivot[str(Q)] = pp_pivot.apply(lambda row: make_cust_sale_grade(row[Q], Q), axis=1)
            
        # 분기별 매출 변화량 feature 생성 + 가중치 부여
        pp_pivot[str(feature_set[i][j])+'_14_2-14_1'] = (pp_pivot['14_2']-pp_pivot['14_1'])*1
        pp_pivot[str(feature_set[i][j])+'_14_3-14_2'] = (pp_pivot['14_3']-pp_pivot['14_2'])*1.1
        pp_pivot[str(feature_set[i][j])+'_14_4-14_3'] = (pp_pivot['14_4']-pp_pivot['14_3'])*1.2
        pp_pivot[str(feature_set[i][j])+'_15_1-14_4'] = (pp_pivot['15_1']-pp_pivot['14_4'])*1.3
        pp_pivot[str(feature_set[i][j])+'_15_2-15_1'] = (pp_pivot['15_2']-pp_pivot['15_1'])*1.4
        pp_pivot[str(feature_set[i][j])+'_15_3-15_2'] = (pp_pivot['15_3']-pp_pivot['15_2'])*1.5
        pp_pivot[str(feature_set[i][j])+'_15_4-15_3'] = (pp_pivot['15_4']-pp_pivot['14_3'])*1.6

        pp_pivot = pp_pivot.drop(columns=Q_list).reset_index()

        # 고객별 분기별 총매출 변동지수(target data) 생성
        if feature_set[i][j] == '구매금액':
            pp_pivot[str(feature_set[i][j])+'_15_3분기_변동지수'] = (pp_pivot[str(feature_set[i][j])+'_14_2-14_1']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_14_3-14_2']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_14_4-14_3']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_15_1-14_4']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_15_2-15_1']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_15_3-15_2'])

            pp_pivot[str(feature_set[i][j])+'_15_4분기_변동지수'] = (pp_pivot[str(feature_set[i][j])+'_14_3-14_2']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_14_4-14_3']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_15_1-14_4']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_15_2-15_1']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_15_3-15_2']) +\
                                    (pp_pivot[str(feature_set[i][j])+'_15_4-15_3'])
        else:
            pass
        
        # 최종 분석 DataFrame 생성(고객 데이터 제외 추후 merge)
        pp_datasets = pd.merge(pp_datasets, pp_pivot, on='고객번호', how='left')
        
# 변동지수(Target 값) 생성 사용자 함수 : 고객별 분기 매출 등급 변화량 지수화
# 변동지수가 0보다 작은 고객(총 고객의 40%)를 1(True)로 설정
def find_RP_by_cus(x):
    if x < 0:
        return 1 # 감소
    else:
        return 0 # 증가
    
pp_datasets['구매금액_15_3분기_변동지수'] = pp_datasets['구매금액_15_3분기_변동지수'].apply(find_RP_by_cus)
pp_datasets['구매금액_15_4분기_변동지수'] = pp_datasets['구매금액_15_4분기_변동지수'].apply(find_RP_by_cus)
pp_datasets.fillna(0)
pp_datasets.rename(index={'분기':'index'})
pp_datasets

고객번호  패션잡화_14_2-14_1  패션잡화_14_3-14_2  패션잡화_14_4-14_3  패션잡화_15_1-14_4  \
0          1             1.0             0.0             0.0             0.0   
1          2             0.0             1.1            -1.2             1.3   
2          3             0.0             0.0             0.0             0.0   
3          4             1.0            -1.1            -2.4             1.3   
4          5             0.0             0.0            -4.8             0.0   
...      ...             ...             ...             ...             ...   
19378  19379             0.0             3.3             1.2             0.0   
19379  19380             0.0             0.0             0.0             0.0   
19380  19381             0.0             0.0             0.0             5.2   
19381  19382             0.0             0.0             0.0             3.9   
19382  19383             0.0             0.0             0.0             0.0   

       패션잡화_15_2-15_1  패션잡화_15_3-15_2  패션잡화_15_4-15_3  식품_14_2-14_1  \
0                 0.0             0.0            -1.6           1.0   
1                -1.4             1.5            -1.6           0.0   
2                 0.0             0.0             3.2           0.0   
3                 0.0             1.5            -1.6           2.0   
4                 0.0             0.0             0.0          -1.0   
...               ...             ...             ...           ...   
19378            -1.4             1.5             3.2           0.0   
19379             4.2            -4.5             6.4           0.0   
19380            -5.6             4.5             3.2           0.0   
19381            -4.2             0.0             0.0           0.0   
19382             0.0             7.5             6.4           0.0   

       식품_14_3-14_2  ...  고가_15_4-15_3  구매금액_14_2-14_1  구매금액_14_3-14_2  \
0              -3.3  ...          -3.2               0             0.0   
1              -1.1  ...          -1.6              -1             1.1   
2               1.1  ...          -1.6               0             0.0   
3              -3.3  ...           0.0               2             0.0   
4              -1.1  ...           3.2              -1             0.0   
...             ...  ...           ...             ...             ...   
19378           0.0  ...           6.4               0             0.0   
19379           0.0  ...           6.4               0             0.0   
19380           0.0  ...           6.4               0             0.0   
19381           0.0  ...           4.8               0             0.0   
19382           0.0  ...           6.4               0             0.0   

       구매금액_14_4-14_3  구매금액_15_1-14_4  구매금액_15_2-15_1  구매금액_15_3-15_2  \
0                 1.2            -1.3             0.0             0.0   
1                -2.4             2.6            -1.4             1.5   
2                -1.2             0.0             0.0             0.0   
3                -1.2             0.0             0.0             0.0   
4                -4.8             5.2            -4.2            -1.5   
...               ...             ...             ...             ...   
19378             2.4             0.0             1.4            -3.0   
19379             0.0             0.0             0.0             0.0   
19380             0.0             2.6             1.4             1.5   
19381             0.0             1.3             1.4             1.5   
19382             0.0             0.0             0.0             3.0   

       구매금액_15_4-15_3  구매금액_15_3분기_변동지수  구매금액_15_4분기_변동지수  
0                -1.6                 1                 1  
1                -1.6                 0                 1  
2                -1.6                 1                 1  
3                -3.2                 0                 1  
4                 1.6                 1                 1  
...               ...               ...               ...  
19378   

# 머신러닝

## 라이브러리 import

In [6]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, roc_auc_score
from sklearn.metrics import f1_score, confusion_matrix, precision_recall_curve, roc_curve
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from lightgbm import LGBMClassifier
from xgboost import XGBClassifier


## 머신러닝 평가 지표 출력 사용자 함수 생성

In [7]:
def get_clf_eval(y_test, pred=None, pred_proba=None):
    confusion = confusion_matrix( y_test, pred)
    accuracy = accuracy_score(y_test , pred)
    precision = precision_score(y_test , pred)
    recall = recall_score(y_test , pred)
    f1 = f1_score(y_test,pred)
    # ROC-AUC 추가 
    roc_auc = roc_auc_score(y_test, pred_proba)
    print('오차 행렬')
    print(confusion)
    # ROC-AUC print 추가
    print('정확도: {0:.4f}, 정밀도: {1:.4f}, 재현율: {2:.4f},\
          F1: {3:.4f}, ROC-'AUC:{4:.4f}'.format(accuracy, precision, recall, f1, roc_auc))

## 머신러닝 사용자 함수

In [29]:
def ml(df_final,pp_demo):
    # 고객 Demo Data -> One-Hot 인코딩
    pp_demo = pd.get_dummies(pp_demo)
    # 고객 Demo Merge
    final_df = pd.merge(pp_demo, df_final, on='고객번호').fillna(0)
    # 최종 학습 데이터
    display(final_df)
    
    # 학슴읠 위해 Feature Data 와 Target 값 분리
    a = pd.Series(final_df.columns, name='컬럼')[pd.Series(final_df.columns, name='컬럼').str.contains('15_3-15_2')].values.tolist()
    b = pd.Series(final_df.columns, name='컬럼')[pd.Series(final_df.columns, name='컬럼').str.contains('15_4-15_3')].values.tolist()
    li = ['구매금액_15_3분기_변동지수','구매금액_15_4분기_변동지수']
    li = li + a + b
    X_train = final_df.drop(columns=li)
    for i in X_train.columns:
        X_train[i]
    y_train = final_df['구매금액_15_3분기_변동지수']
    
    a = pd.Series(final_df.columns, name='컬럼')[pd.Series(final_df.columns, name='컬럼').str.contains('14_2-14_1')].values.tolist()
    b = pd.Series(final_df.columns, name='컬럼')[pd.Series(final_df.columns, name='컬럼').str.contains('15_4-15_3')].values.tolist()
    li = ['구매금액_15_3분기_변동지수','구매금액_15_4분기_변동지수']
    li = li + a + b
    X_test = final_df.drop(columns=li)
    y_test = final_df['구매금액_15_4분기_변동지수']


    # 로지스틱, 결정트리. 랜덤포레스트, LightGBM, XGboost로 학습,예측 및 평가 수행. 
    lr_clf = LogisticRegression(solver='liblinear')
    dt_clf = DecisionTreeClassifier()
    rf_clf = RandomForestClassifier()
    lgbm_clf = LGBMClassifier(n_estimators=1000, num_leaves=50, n_jobs=-1, boost_from_average=False)
    xgb_clf = XGBClassifier(n_estimators=1000, n_jobs=-1 )
    
    lr_clf.fit(X_train , y_train)
    dt_clf.fit(X_train , y_train)
    rf_clf.fit(X_train , y_train)
    lgbm_clf.fit(X_train, y_train)
    xgb_clf.fit(X_train, y_train)
    
    pred_lr = lr_clf.predict(X_test)
    pred_dt = dt_clf.predict(X_test)
    pred_rf = rf_clf.predict(X_test)
    pred_lgbm = lgbm_clf.predict(X_test)
    pred_xgb = xgb_clf.predict(X_test)
    
    pred_proba_lr = lr_clf.predict_proba(X_test)[:, 1]
    pred_proba_dt = dt_clf.predict_proba(X_test)[:, 1]
    pred_proba_rf = rf_clf.predict_proba(X_test)[:, 1]
    pred_proba_lgbm = lgbm_clf.predict_proba(X_test)[:, 1]
    pred_proba_xgb = xgb_clf.predict_proba(X_test)[:, 1]
    
    get_clf_eval(y_test, pred_lr, pred_proba_lr)
    get_clf_eval(y_test, pred_dt, pred_proba_dt)
    get_clf_eval(y_test, pred_rf, pred_proba_rf)
    get_clf_eval(y_test, pred_lgbm, pred_proba_lgbm)
    get_clf_eval(y_test, pred_xgb, pred_proba_xgb)
    
    return pred_lr
pred_lr = ml(pp_datasets,pp_demo)
pred_lr

고객번호  성별_F  성별_M  연령대_19세이하  연령대_20세~24세  연령대_25세~29세  연령대_30세~34세  \
0          1     0     1          0            0            0            0   
1          2     0     1          0            0            0            0   
2          3     0     1          0            0            0            0   
3          4     1     0          0            0            0            0   
4          5     0     1          0            0            0            0   
...      ...   ...   ...        ...          ...          ...          ...   
19378  19379     1     0          0            1            0            0   
19379  19380     1     0          0            1            0            0   
19380  19381     1     0          0            0            0            0   
19381  19382     1     0          0            0            0            0   
19382  19383     1     0          0            0            1            0   

       연령대_35세~39세  연령대_40세~44세  연령대_45세~49세  ...  고가_15_4-15_3  \
0                0            0            0  ...          -3.2   
1                0            0            0  ...          -1.6   
2                0            0            0  ...          -1.6   
3                0            0            0  ...           0.0   
4                0            0            0  ...           3.2   
...            ...          ...          ...  ...           ...   
19378            0            0            0  ...           6.4   
19379            0            0            0  ...           6.4   
19380            0            1            0  ...           6.4   
19381            0            0            1  ...           4.8   
19382            0            0            0  ...           6.4   

       구매금액_14_2-14_1  구매금액_14_3-14_2  구매금액_14_4-14_3  구매금액_15_1-14_4  \
0                   0             0.0             1.2            -1.3   
1                  -1             1.1            -2.4             2.6   
2                   0             0.0            -1.2             0.0   
3                   2             0.0            -1.2             0.0   
4                  -1             0.0            -4.8             5.2   
...               ...             ...             ...             ...   
19378               0             0.0             2.4             0.0   
19379               0             0.0             0.0             0.0   
19380               0             0.0             0.0             2.6   
19381               0             0.0             0.0             1.3   
19382               0             0.0             0.0             0.0   

       구매금액_15_2-15_1  구매금액_15_3-15_2  구매금액_15_4-15_3  구매금액_15_3분기_변동지수  \
0                 0.0             0.0            -1.6                 1   
1                -1.4             1.5            -1.6                 0   
2                 0.0             0.0            -1.6                 1   
3                 0.0             0.0            -3.2                 0   
4                -4.2            -1.5             1.6                 1   
...               ...             ...             ...               ...   
19378             1.4            -3.0             4.8                 0   
19379             0.0             0.0             6.4                 0   
19380             1.4             1.5             4.8                 0   
19381             1.4             1.5             4.8                 0   
19382             0.0             3.0             4.8                 0   

       구매금액_15_4분기_변동지수  
0                     1  
1                     1  
2                     1  
3                     1  
4                     1  
...                 ...  
19378                 0  
19379                 0  
19380                 0  
19381                 0  
19382                 0  

[19383 rows x 112 columns]

[11:54:52] WARNING: ../src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
오차 행렬
[[7852 2250]
 [3089 6192]]
정확도: 0.7246, 정밀도: 0.7335, 재현율: 0.6672,          F1: 0.6988, AUC:0.8072
오차 행렬
[[6148 3954]
 [3945 5336]]
정확도: 0.5925, 정밀도: 0.5744, 재현율: 0.5749,          F1: 0.5747, AUC:0.5918
오차 행렬
[[7541 2561]
 [3838 5443]]
정확도: 0.6699, 정밀도: 0.6800, 재현율: 0.5865,          F1: 0.6298, AUC:0.7409
오차 행렬
[[7498 2604]
 [3187 6094]]
정확도: 0.7012, 정밀도: 0.7006, 재현율: 0.6566,          F1: 0.6779, AUC:0.7660
오차 행렬
[[7292 2810]
 [3300 5981]]
정확도: 0.6848, 정밀도: 0.6804, 재현율: 0.6444,          F1: 0.6619, AUC:0.7515


array([1, 1, 1, ..., 0, 0, 0])

## '로지스틱 모델를 최종 모델로 선정'

# 추천 시스템

## Target 고객의 고객번호
- 총 고객 19383 중 8442명 선정

In [30]:
# 추천 시스템용 데이터(recommend)
reco = pd.merge(pp_demo, pp_datasets, on='고객번호', how='left')
reco_cust = reco[pred_lr==1]
# 마케팅이 필요하다고 판단되는 고객번호 배열 생성
reco_cust_num = reco_cust['고객번호'].values
reco_cust_num,len(reco_cust_num)

(array([    1,     2,     3, ..., 19346, 19360, 19364]), 8442)

## 추천을 위한 종합 특성 DataFrame 생성

In [10]:
pp_reco = pd.merge(pp_purprd_prodcl, pp_demo, on='고객번호', how='left')
pp_reco_df = pp_reco[pp_reco['분기']!='15_4']
pp_reco_df

고객번호 제휴사    영수증번호  점포코드  구매_연  구매_월  구매_일  구매시간    소분류코드   구매금액  \
0             1   A  1182697    12  2014     1     8    14  A010302   4800   
1             1   A  1239229    12  2014     1    18    15  A010302   3500   
2             1   A  1307657    12  2014     1    30    11  A010302   3900   
3             1   A  1402358    12  2014     2    18    16  A010302   3500   
4             1   A  1456379    12  2014     2    28    15  A010302   3500   
...         ...  ..      ...   ...   ...   ...   ...   ...      ...    ...   
28437049  18697   C   631324    29  2014     5    30    21  C060405   2300   
28437050  18697   C   631629    29  2014     7    20    15  C060405   2300   
28437051  18710   C  1073027    75  2015     1    22    12  C030312   7990   
28437052  18710   C  1073138    75  2015     1    29    21  C030312   7990   
28437053  19364   D   209812    13  2015     7    11    21  D040112  10500   

          ... 대중분류명  중분류명 제휴사_중분류명    소분류명 TYPE lev    분기 성별      연령대  \
0         ...   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  M    60세이상   
1         ...   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  M    60세이상   
2         ...   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  M    60세이상   
3         ...   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  M    60세이상   
4         ...   유제품  축산가공    A축산가공     유제품  편의품  중가  14_1  M    60세이상   
...       ...   ...   ...      ...     ...  ...  ..   ... ..      ...   
28437049  ...  즉석식품  베이커리    C베이커리   페스츄리류  편의품  중가  14_2  M  50세~54세   
28437050  ...  즉석식품  베이커리    C베이커리   페스츄리류  편의품  중가  14_3  M  50세~54세   
28437051  ...   채소류   근채류     C근채류  기타한방약재  편의품  고가  15_1  F  50세~54세   
28437052  ...   채소류   근채류     C근채류  기타한방약재  편의품  고가  15_1  F  50세~54세   
28437053  ...  공산식품    과자      D과자  컵아이스크림  편의품  중가  15_3  F  25세~29세   

               거주지역  
0         서울특별시 강남구  
1         서울특별시 강남구  
2         서울특별시 강남구  
3         서울특별시 강남구  
4         서울특별시 강남구  
...             ...  
28437049        경기도  
28437050        경기도  
28437051       전라북도  
28437052       전라북도  
28437053      인천광역시  

[28437054 rows x 21 columns]

In [33]:
# pp_reco_df = pp_reco[pp_reco['분기']!='15_4']
# pp_reco_df = pp_reco_df[['소분류명', '고객번호', '성별', '연령대', '거주지역','중분류명']]  # 25006807
# pp_reco_df = pp_reco_df.drop_duplicates(['소분류명', '고객번호'], keep='first').reset_index(drop=True)
# pp_reco_df = pp_reco_df.drop(columns=['고객번호'])
# pp_reco_df = pp_reco_df.drop_duplicates(['소분류명', '성별', '연령대', '거주지역', '중분류명'], keep='first').reset_index(drop=True)
# pp_reco_df['연령대'] = pp_reco_df['연령대'].apply(lambda x:x[:2])
# # pp_reco_df = pp_reco_df[(pp_reco_df['소분류명']=='유제품') & (pp_reco_df['성별']=='M')]
# pp_reco_df

## 고객별 제휴사_중분류명별 구매 횟수를 사용한 선호도 점수화

In [14]:
# p for customer: 수정 필요 ( 제휴사와 채널 정보 추가 )

In [53]:
reco_cust_pv = pp_purprd_prodcl.pivot_table(index=['제휴사_중분류명'],columns='고객번호',values='구매금액',aggfunc='count')
reco_cust_pv = reco_cust_pv.unstack().dropna().reset_index()
reco_cust_pv = reco_cust_pv.rename(columns={0:'count'})
reco_cust_pv

고객번호 제휴사_중분류명  count
0            1    A가공식품  157.0
1            1    A건강식품    9.0
2            1    A골프용품   13.0
3            1      A구두    9.0
4            1  A남성 트랜디    4.0
...        ...      ...    ...
3105184  19383    D집중케어   15.0
3105185  19383     D클렌징    6.0
3105186  19383  D핸드/풋케어    8.0
3105187  19383    D헤어케어   12.0
3105188  19383  D화장지/티슈    5.0

[3105189 rows x 3 columns]

## count 값을 로그값을 취한 후 0~5 사이로 표준화

In [44]:
max_cnt = reco_cust_pv.describe().loc['max','count']
max_cnt

1324.0

In [60]:
reco_cust_pv['count'] = reco_cust_pv['count'].apply(lambda x: np.log1p(x))
max_cnt = reco_cust_pv.describe().loc['max','count']
reco_cust_pv['count'] = reco_cust_pv['count'].apply(lambda x: 5*x/max_cnt)

reco_cust_pv

고객번호 제휴사_중분류명     count
0            1    A가공식품  4.645950
1            1    A건강식품  3.755145
2            1    A골프용품  3.917915
3            1      A구두  3.755145
4            1  A남성 트랜디  3.315251
...        ...      ...       ...
3105184  19383    D집중케어  3.976035
3105185  19383     D클렌징  3.550367
3105186  19383  D핸드/풋케어  3.698561
3105187  19383    D헤어케어  3.884181
3105188  19383  D화장지/티슈  3.448612

[3105189 rows x 3 columns]

In [61]:
from surprise.model_selection import cross_validate
from surprise import Reader, Dataset
from surprise import SVD
from surprise import Dataset
from surprise import accuracy
from surprise.model_selection import train_test_split

# 판다스 DataFrame에서 Surprise 데이터 세트로 데이터 로딩
reader = Reader(rating_scale=(0, 5.0))
data = Dataset.load_from_df(reco_cust_pv, reader)

algo = SVD(random_state=0)
cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=2, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 2 split(s).

                  Fold 1  Fold 2  Mean    Std     
RMSE (testset)    0.5305  0.5299  0.5302  0.0003  
MAE (testset)     0.4374  0.4367  0.4370  0.0004  
Fit time          13.86   13.64   13.75   0.11    
Test time         10.26   10.94   10.60   0.34    


{'test_rmse': array([0.53051438, 0.52985352]),
 'test_mae': array([0.43739943, 0.43666302]),
 'fit_time': (13.859685897827148, 13.63912320137024),
 'test_time': (10.258793115615845, 10.938879013061523)}

In [72]:
from surprise.model_selection import GridSearchCV
# 최적 파라미터는 {'n_epochs':[10], 'n_factors':[50]} 지만 너무 오래 걸려서 못 함.
# 최적화할 파라미터를 딕셔너리 형태로 지정
param_grid = {'n_epochs':[30], 'n_factors':[50]}# n_epochs : 점진적 하강 방식의 반복 횟수, n_factors : 잠재요인 크기 K   mxn = mxp + nxp K 는 p의 크기 

# CV를 3개 폴드 세트로 지정, 성능 평가는 rmse, mse로 수행하도록 GridSearchCV 구성
gs = GridSearchCV(SVD, param_grid, measures=['rmse', 'mae'], cv=2, refit=True)
gs
gs.fit(data)


# 최고 RMSE Evaluation 점수와 그때의 하이퍼 파라미터
print('RMSE : ',gs.best_score['rmse'])
print('최적 파라미터 : ',gs.best_params['rmse'])

RMSE :  0.5162931226582046
최적 파라미터 :  {'n_epochs': 30, 'n_factors': 50}


In [73]:
from surprise.dataset import DatasetAutoFolds

reco_cust_pv.to_csv('./reco_cust_pv_noh.csv', index=False, header=False)

reader = Reader(line_format='user item rating', sep=',', rating_scale=(0, 5.0))
# DatasetAutoFolds 클래스를 ratings_noh.csv 파일 기반으로 생성
data_folds = DatasetAutoFolds(ratings_file='./reco_cust_pv_noh.csv', reader=reader)

# 전체 데이터를 학습 데이터로 생성함.
trainset = data_folds.build_full_trainset()

In [74]:
algo = SVD(n_epochs=30, n_factors=50, random_state=0)
algo.fit(trainset)


In [26]:
pp_reco_df['제휴사_중분류명'].unique()

array(['유제품', '청과', '멸치류', ..., '기타탁구용품', '여성발가락', '기타한방약재'], dtype=object)

In [84]:
def recomm_item_by_surprise(algo, userId, top_n=10):
    
    # 알고리즘 객체의 predict() 메서드를 소분류명에 반복 수행한 후 결과를 list 객체로 저장
    predictions = [algo.predict(str(userId), str(s_cat)) for s_cat in pp_reco_df['제휴사_중분류명'].unique()]
    
    # predictions list 객체는 surprise의 Predictions 객체를 원소로 가지고 있음
    # [Prediction(uid='9'm iid='1', est=3.69)....]
    
    # 이를 est 값으로 정렬하기 위해서 아래의 sortkey_est 함수를 정의함
    # sortkey_set 함수는 list 객체의 sort() 함수의 키 값으로 사용되어 정렬 수행.
    def sortkey_est(pred):
        return pred.est
    
    # sortket_est() 반환값의 내림 차순으로 정렬 수행하고 top_n 개의 최상위 값 추출.
    predictions.sort(key=sortkey_est, reverse=True)
    top_predictions = predictions[:top_n]
    display(top_predictions)
    
    # top_n으로 추출된 소분류명의 정보 추출, 아이템 아이디, 추천 예상 평점
    top_item_ids = [ str(pred.iid) for pred in top_predictions]
    top_item_rating = [ pred.est for pred in top_predictions]
    top_item_titles = pp_reco_df[pp_reco_df.제휴사_중분류명.isin(top_item_ids)]['제휴사_중분류명'].unique()
    # display(top_item_ids)
    top_item_preds = [ (id, title, rating) for id, title, rating in zip(top_item_ids, top_item_titles, top_item_rating)]
    
    return top_item_preds

# unseen_movies = get_unseen_surprise(ratings, movies, 9)
top_item_preds = recomm_item_by_surprise(algo, 5500, top_n=10)

print(' Top 10 추천 아이템 리스트')
for top_item in top_item_preds:
    print(top_item[1], ':', top_item[2])

[Prediction(uid='5500', iid='B라면', r_ui=None, est=4.543046730526903, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B스낵', r_ui=None, est=4.494923205948669, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B흰우유', r_ui=None, est=4.443539816791388, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B두부', r_ui=None, est=4.372380759985807, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B비스켓', r_ui=None, est=4.323179194747924, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B잎채소', r_ui=None, est=4.265218712410821, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B요구르트', r_ui=None, est=4.263149139480337, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B어묵맛살', r_ui=None, est=4.227923412222189, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B열매채소', r_ui=None, est=4.219599711726999, details={'was_impossible': False}),
 Prediction(uid='5500', iid='B계란', r_ui=None, 

 Top 10 추천 아이템 리스트
B흰우유 : 4.543046730526903
B요구르트 : 4.494923205948669
B계란 : 4.443539816791388
B라면 : 4.372380759985807
B두부 : 4.323179194747924
B어묵맛살 : 4.265218712410821
B열매채소 : 4.263149139480337
B잎채소 : 4.227923412222189
B스낵 : 4.219599711726999
B비스켓 : 4.205972095685121


In [85]:
reco_cust_pv[reco_cust_pv['고객번호']==5500].sort_values(by='count',ascending=False)

고객번호 제휴사_중분류명     count
854734  5500    B탄산음료  4.626521
854652  5500      B스낵  4.554740
854593  5500      B라면  4.509954
854590  5500      B두부  4.418463
854621  5500     B비스켓  4.409353
...      ...      ...       ...
854635  5500      B선식  2.265555
854639  5500   B성인침구류  2.265555
854709  5500   B조명/전구  2.265555
854708  5500    B조리용기  2.265555
854775  5500    C탄산음료  2.265555

[242 rows x 3 columns]

In [86]:
uid = str(5500)
iid = str('A가공식품')

pred = algo.predict(uid, iid, verbose=True)

user: 5500       item: A가공식품      r_ui = None   est = 2.86   {'was_impossible': False}
